src/System.IO.Pipes.AccessControl/src/System/IO/PipesAclExtensions.cs

```diff
@@ -12,28 +12,31 @@ public static class PipesAclExtensions
         [System.Security.SecurityCritical]
         public static PipeSecurity GetAccessControl(this PipeStream stream)
         {
-            // PipeState can not be Closed and the Handle can not be null or closed
-            var handle = stream.SafePipeHandle; // A non-null, open handle implies a non-closed PipeState.
+            // Checks that State != WaitingToConnect and State != Closed
+            var handle = stream.SafePipeHandle;
+
+            // PipeState must be Disconnected, Connected, or Broken
             return new PipeSecurity(handle, AccessControlSections.Access | AccessControlSections.Owner | AccessControlSections.Group);
         }
 
         [System.Security.SecurityCritical]
         public static void SetAccessControl(this PipeStream stream, PipeSecurity pipeSecurity)
         {
-            // PipeState can not be Closed and the Handle can not be null or closed
             if (pipeSecurity == null)
             {
                 throw new ArgumentNullException(nameof(pipeSecurity));
             }
 
-            var handle = stream.SafePipeHandle; // A non-null, open handle implies a non-closed PipeState.
+            // Checks that State != WaitingToConnect and State != Closed
+            var handle = stream.SafePipeHandle;
 
-            // NamedPipeClientStream: PipeState can not be WaitingToConnect or Broken. WaitingToConnect is covered by the SafePipeHandle check.
-            if (stream is NamedPipeClientStream && !stream.IsConnected) // Pipe could also be WaitingToConnect, Broken, Disconnected, or Closed.
+            // Checks that State != Broken
+            if (stream is NamedPipeClientStream && !stream.IsConnected)
             {
-                throw new InvalidOperationException(SR.InvalidOperation_PipeNotYetConnected);
+                throw new IOException(SR.IO_IO_PipeBroken);
```

Comment:
@weshaggard the reason I was able to change our InvalidOperationException to IOException in the SetAccessControl check specific to NamedPipeClientStream is that the stream.SafePipeHandle check is effectively validating for us that the PipeState != WaitingToConnect. That means we don't need to check for it explicitly in the NamedPipeClientStream check, so we know that if it isn't connected then it can only be Broken.